File: src/apple/Packages/MapleUI/Sources/MapleUI/Molecules/MuiFrameTimeHUD.swift
// MuiFrameTimeHUD.swift — Maple UI Molecules-L1 (unified-component-catalog.md
// §2.3). Performance readout overlay, built from Text only. Color-codes
// against Maple's own perf invariants (16ms slider-tick target, 50ms hard
// limit — CONTRIBUTING.md "Performance invariants") so the HUD itself flags
// a budget breach rather than just reporting a number.

import SwiftUI

public enum MuiFrameTimeStatus: Sendable {
    case good, warn, bad
}

public struct MuiFrameTimeHUD: View {
    public let frameMs: Double
    public let fps: Int?
    /// Target frame budget in ms — at/under this is `good`.
    public let budgetMs: Double
    /// Hard limit in ms — over this is `bad`; between budget and this is
    /// `warn`.
    public let hardLimitMs: Double

    public init(frameMs: Double, fps: Int? = nil, budgetMs: Double = 16, hardLimitMs: Double = 50) {
        self.frameMs = frameMs
        self.fps = fps
        self.budgetMs = budgetMs
        self.hardLimitMs = hardLimitMs
    }

    public var body: some View {
        let status = Self.status(frameMs: frameMs, budgetMs: budgetMs, hardLimitMs: hardLimitMs)
        MuiText(
            "\(String(format: "%.1f", frameMs))ms avg · \(Self.displayFps(frameMs: frameMs, fps: fps))fps",
            variant: .valueChip,
            color: .main
        )
        .padding(.horizontal, MuiTokens.spacingSm)
        .padding(.vertical, 4)
        .background(Self.color(for: status).opacity(0.18), in: Capsule())
        .overlay(Capsule().stroke(Self.color(for: status), lineWidth: 1))
        .accessibilityElement(children: .combine)
    }

    /// The rounded frames-per-second readout — `fps` when explicitly
    /// supplied, otherwise derived from `frameMs`. Public + static so this
    /// is unit-testable without rendering a view.
    public static func displayFps(frameMs: Double, fps: Int?) -> Int {
        if let fps { return fps }
        guard frameMs > 0 else { return 0 }
        return Int((1000 / frameMs).rounded())
    }

    /// Classifies `frameMs` against the budget/hard-limit thresholds.
    /// Public + static so this is unit-testable without rendering a view.
    public static func status(frameMs: Double, budgetMs: Double, hardLimitMs: Double) -> MuiFrameTimeStatus {
        if frameMs > hardLimitMs { return .bad }
        if frameMs > budgetMs { return .warn }
        return .good
    }

    private static func color(for status: MuiFrameTimeStatus) -> Color {
        switch status {
        case .good: return MuiTokens.successText
        case .warn: return MuiTokens.warn
        case .bad: return MuiTokens.errorText
        }
    }
}

#Preview("MuiFrameTimeHUD") {
    VStack(spacing: 12) {
        MuiFrameTimeHUD(frameMs: 11.2)
        MuiFrameTimeHUD(frameMs: 28.4)
        MuiFrameTimeHUD(frameMs: 62.1)
    }
    .padding()
    .background(MuiTokens.imageCanvas)
}

File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiFrameTimeHud.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Frame-time HUD molecule (unified-component-catalog.md
    /// §2.3, "Frame-time HUD" row: "Performance readout overlay", built
    /// from Text only) — color-codes against Maple's own perf invariants
    /// (CLAUDE.md: 16ms slider-tick target, 50ms hard limit) via
    /// <see cref="MuiFrameTimeClassifier"/>, so the HUD flags a budget
    /// breach rather than just reporting a bare number.
    ///
    /// Ports `mui-frame-time-hud.component.ts` 1:1, driven entirely by the
    /// shared, unit-tested classifier rather than duplicating the
    /// good/warn/bad thresholds here.
    /// </summary>
    public sealed class MuiFrameTimeHud : ContentControl
    {
        public static readonly DependencyProperty FrameMsProperty =
            DependencyProperty.Register(nameof(FrameMs), typeof(double), typeof(MuiFrameTimeHud),
                new PropertyMetadata(0.0, (d, _) => ((MuiFrameTimeHud)d).Rebuild()));

        public static readonly DependencyProperty FpsProperty =
            DependencyProperty.Register(nameof(Fps), typeof(double?), typeof(MuiFrameTimeHud),
                new PropertyMetadata(null, (d, _) => ((MuiFrameTimeHud)d).Rebuild()));

        public static readonly DependencyProperty BudgetMsProperty =
            DependencyProperty.Register(nameof(BudgetMs), typeof(double), typeof(MuiFrameTimeHud),
                new PropertyMetadata(MuiFrameTimeClassifier.DefaultBudgetMs, (d, _) => ((MuiFrameTimeHud)d).Rebuild()));

        public static readonly DependencyProperty HardLimitMsProperty =
            DependencyProperty.Register(nameof(HardLimitMs), typeof(double), typeof(MuiFrameTimeHud),
                new PropertyMetadata(MuiFrameTimeClassifier.DefaultHardLimitMs, (d, _) => ((MuiFrameTimeHud)d).Rebuild()));

        public double FrameMs
        {
            get => (double)GetValue(FrameMsProperty);
            set => SetValue(FrameMsProperty, value);
        }

        /// <summary>Defaults to 1000 / FrameMs, rounded, when unset.</summary>
        public double? Fps
        {
            get => (double?)GetValue(FpsProperty);
            set => SetValue(FpsProperty, value);
        }

        /// <summary>Target frame budget in ms — at/under this is Good.</summary>
        public double BudgetMs
        {
            get => (double)GetValue(BudgetMsProperty);
            set => SetValue(BudgetMsProperty, value);
        }

        /// <summary>Hard limit in ms — over this is Bad; between budget and
        /// this is Warn.</summary>
        public double HardLimitMs
        {
            get => (double)GetValue(HardLimitMsProperty);
            set => SetValue(HardLimitMsProperty, value);
        }

        private readonly StackPanel _root = new() { Orientation = Orientation.Horizontal, Spacing = 8 };
        private readonly MuiText _fpsText = new() { Variant = MuiTextVariant.ValueChip };
        private readonly MuiText _frameMsText = new() { Variant = MuiTextVariant.ValueChip };

        public MuiFrameTimeHud()
        {
            _root.Children.Add(_fpsText);
            _root.Children.Add(_frameMsText);
            Content = _root;
            IsTabStop = false;
            IsHitTestVisible = false;

            Rebuild();
        }

        private void Rebuild()
        {
            var fps = MuiFrameTimeClassifier.ComputeFps(FrameMs, Fps);
            var status = MuiFrameTimeClassifier.Classify(FrameMs, BudgetMs, HardLimitMs);
            var colorRole = status switch
            {
                MuiFrameTimeStatus.Bad => MuiTextColorRole.Error,
                MuiFrameTimeStatus.Warn => MuiTextColorRole.Warning,
                _ => MuiTextColorRole.Success,
            };

            _fpsText.Text = $"{fps:0} fps";
            _fpsText.ColorRole = colorRole;

            _frameMsText.Text = $"{FrameMs:0.0} ms";
            _frameMsText.ColorRole = colorRole;

            var statusName = status switch
            {
                MuiFrameTimeStatus.Bad => "over budget",
                MuiFrameTimeStatus.Warn => "near budget",
                _ => "within budget",
            };
            AutomationProperties.SetName(this, $"{fps:0} frames per second, {FrameMs:0.0} milliseconds, {statusName}");
        }
    }
}
